applications/tari_merge_mining_proxy/src/proxy.rs

```diff
@@ -317,75 +320,77 @@ impl InnerService {
+                        "Block `{}` submitted but no matching block template was found, possible duplicate submission",
+                        hex::encode(&hash)
+                    );
+                    continue;
                 },
             };
 
-            debug!(target: LOG_TARGET, "Located Tari Hash: {:?}", hash);
-
-            if let Some(mut block_data) = self.block_templates.get(&hash).await {
-                let monero_data = helpers::construct_monero_data(monero_block, block_data.clone().monero_seed)?;
-
-                let pow_data = bincode::serialize(&monero_data)?;
-                let h = block_data.tari_block.header.as_mut().unwrap();
-                let height = h.height;
-                let p = h.pow.as_mut().unwrap();
-                p.pow_data = pow_data;
-
-                let mut base_node_client = self.connect_grpc_client().await?;
-                let mut block_templates_clone = self.block_templates.clone();
-                let block_data_clone = block_data.clone();
-                handle.spawn(async move {
-                    let start = std::time::Instant::now();
-                    match base_node_client
-                        .submit_block(block_data_clone.tari_block)
-                        .await
-                        .map_err(|status| MmProxyError::GrpcRequestError {
-                            status,
-                            details: "failed to submit block".to_string(),
-                        }) {
-                        Ok(..) => {
-                            debug!(
-                                target: LOG_TARGET,
-                                "Submitted block #{} to Tari node in {:.0?} (SubmitBlock)",
-                                height,
-                                start.elapsed()
-                            );
-                            block_templates_clone.remove(&hash).await;
-                        },
-                        _ => debug!(
-                            target: LOG_TARGET,
-                            "Problem submitting block #{} to Tari node, responded in  {:.0?} (SubmitBlock)",
-                            height,
-                            start.elapsed()
-                        ),
+            let monero_data = helpers::construct_monero_data(monero_block, block_data.monero_seed.clone())?;
+
+            let header_mut = block_data.tari_block.header.as_mut().unwrap();
+            let height = header_mut.height;
+            header_mut.pow.as_mut().unwrap().pow_data = bincode::serialize(&monero_data)?;
+
+            let mut base_node_client = self.connect_grpc_client().await?;
+            let start = Instant::now();
+            match base_node_client.submit_block(block_data.tari_block).await {
+                Ok(_) => {
+                    json_resp["result"]["tari_status"] = json!("OK");
+
+                    // If monero returned an error, remove it
+                    if !json_resp["error"].is_null() {
```

Comment:
Return `accept_response` here, it is in the format which xmrig expects. See line 644 on how to construct it. Was going to embed a message into the json as well to indicate that it was accepted to Tari which we could optionally display in XMRig.

Reply:
The json rpc format is still adhered to (by monerod), we are just changing the contents of the monero response. Happy to change it to create a new response (getting rid of the `"error": null` field at least)

Reply:
I haven't tested it but I think `"error": null` could potentially cause problems for some applications. 

Anything that checks for the "presence of" and doesn't necessarily check the null (as it is not necessarily expected here) and tries to access as normal, will result in a segfault.

While this is actually a problem from the side of the consumer application, it brings with the complaint of "But it works everywhere else except here" from a user's point of view.

Thoughts?

Reply:
Yeah, I've removed `"error": null`.

I've just noticed "special code" for submit_block in the proxy_request_to_monerod function. We probably want to consolidate functionality to the handler for submit_block. Will update on this PR with what I've changed

Reply:
The "special code" you're referring to is probably related to using XMRig in self-select mode.

Reply:
I've kept the `submit_block` and `proxy_submit_to_origin == false` special case as it is not within the scope of this PR to change anything there. 

In the case where a pool is using the Tari merge mining proxy, the behaviour is as follows:
Block is submitted to monero (if proxy_submit_to_origin == true)
Block is submitted to tari
if monero and/or tari succeeds, return a success result
if monero and tari fail, return an error result
if proxy_submit_to_origin == false, then we need to fail if tari fails (even though the response is success for monerod because of aforementioned config)